### ISSUE TITLE
[Feature] - add ARM target

### ISSUE THREAD

**Issue by vmakwana** [nexe/nexe#424]
Used nexe on macOS, compiled, transferred the built binary (test) to raspberry pi, ran ./test and the following showed up: -bash: ./test: cannot execute binary file: Exec format error

**Comment by jaredallard** [comment]
@vmakwana Raspberry Pi is ARM, from the error you're getting you're trying to run a x86 binary on it

**Comment by calebboyd** [comment]
Hmm. To start, would have to add tests for `linux-arm-<NODE_VERSION>`
 - test/target.spec.ts -- is entries of `[input, expected]`

Add, mappings to `arm71` in this hash: 
https://github.com/nexe/nexe/blob/23f550818bba51253f3120a5301d3c27e67de5f3/src/target.ts#L29
(eg arm, arm7, arm71 all map to arm71)
could also do `arm6: 'arm64'`

The more complicated part is the docker build in CI -- 
Could rename`tasks/docker.ts` to `tasks/docker-apline.ts` and add `docker-arm` with appropriate contents... would also have to adjust the `asset-compile.ts` task

**Comment by jaredallard** [comment]
@calebboyd I'll try to take a crack at this over the holiday weekend if time permits

**Comment by calebboyd** [comment]
@jaredallard do you have anything on this? If not I'll push a branch later to collab on

**Comment by jaredallard** [comment]
@calebboyd Sorry for the delayed response... Email client messed up and never got past the outbox. :/

No I wasn't able to work on it this weekend, had a urgent work project over the weekend, I'd be happy to hop on this (especially as a collab) asap.

**Comment by calebboyd** [comment]
@jaredallard 
https://github.com/nexe/nexe/compare/arm

This will make this function https://github.com/nexe/nexe/blob/master/tasks/asset-compile.ts#L35
contain all the arm71 linux node-version combinations.

Just need to handle them

**Comment by xxxtonixxx** [comment]
any news on this?

**Comment by calebboyd** [comment]
[Nothing yet](https://twitter.com/slicknet/status/782274190451671040). Prioritizing #418

**Comment by pollarize** [comment]
Hi all i have an issue with passing the arguments to node application only on arm64 architecture. The compilation ot 32 bit is working properly. I take a look around with similar problems but i cant figure out way its happen. Do you have any idea how to prevent by passing the argument to node CLI ?

**Comment by calebboyd** [comment]
@pollarize Can you ensure NODE_OPTIONS isn't set?

**Comment by pollarize** [comment]
Yes, i pass arguments to my scritp as follow : node <filename>.js <args>
After compilation i pass arguments: Binary <args> 
Oneach other arhitecture working ass expected only on arm64.
I printed the process.argv array and it's something like  [0] = Binary , [1] = <filename>.js (without giving any arguments)

**Comment by calebboyd** [comment]
@pollarize Try the latest version -- looks like it had broken in a recent release

**Comment by Codelica** [comment]
Is this still being considered?

**Comment by calebboyd** [comment]
Sure! Just haven't gotten a round tuit.

**Comment by defusioner** [comment]
Hello, how could I test my implementation of the docker part locally? Ty

**Comment by calebboyd** [comment]
On an arm platform , docker or otherwise

**Comment by defusioner** [comment]
@calebboyd Is it normal that I have a different behaviour on Docker for Mac (I'm testing basic script) comparing, probably, to Azure runner? 

For example, I've got `curl not found`, now I have `make not found`. What would you recommend me to test my docker file? And could you tell me what does it do after downloading?
```
RUN curl -sSL https://nodejs.org/dist/v${NODE_VERSION}/node-v${NODE_VERSION}.tar.gz | tar -xz && 
  mkdir /nexe_temp && mv /node-v${NODE_VERSION} /nexe_temp/${NODE_VERSION} && 
  cd /nexe_temp/${NODE_VERSION} && 
  ./configure --prefix=/usr --fully-static && 
  make && make install && 
  paxctl -cm /usr/bin/node

RUN rm /nexe_temp/${NODE_VERSION}/out/Release/node && 
  npm install -g nexe@${NEXE_VERSION}
RUN echo "console.log('hello world')" >> index.js && 
  nexe --build --no-mangle --temp /nexe_temp -c="--fully-static" -o out
```

**Comment by megastef** [comment]
If you need arm machines, check http://scaleway.com

**Comment by wilmardo** [comment]
Drone CI has native arm64 hardware and a integrates nicely with Github. See my mosquitto Docker build for example:

https://github.com/LANsible/docker-mosquitto

**Comment by wilmardo** [comment]
Ah I see the needed code is already added?

https://github.com/nexe/nexe/blob/master/src/releases.ts#L53

@calebboyd Is it possible to push the ARM releases to Github? It would help immensely since ARM hardware isn't that powerful so compiling Node each time takes forever :)

**Comment by mturnshek** [comment]
Is it possible to use this now? If so, how? @wilmardo

**Comment by wilmardo** [comment]
@mturnshek The are no ARM releases on Github so not yet without the `--build`
But I have nexe working on ARM like this:

```
nexe  --build  --output zigbee2mqtt
```
https://github.com/LANsible/docker-zigbee2mqtt/blob/master/Dockerfile#L34

**Comment by mturnshek** [comment]
@wilmardo 

Here's the error I get on my RPi 3 B+, which is `arm` architecture (not `arm64`).

The command:
`nexe example.js --build`

```
ℹ nexe 3.3.2
✔ Already downloaded...
✔ Configuring node build: --dest-cpu=arm7l
✔ Finished in 5.007s

Error: python ./configure.py --dest-cpu=arm7l exited with code: 2

See nexe -h for usage..
```

If I try your command, or for that matter anything where the beginning is `nexe  --build`, I get the following error:

```
Error: Entry file "" not found!

See nexe -h for usage..
```

However, I also get that same error for `nexe --build` commands on MacOS. On MacOS, `nexe example.js --build` does compile node and make a binary.

**Comment by wilmardo** [comment]
@mturnshek Might be nicer to split this to a separate issue but for the sake of all the other who might struggle.

The first command should be allright but I suspect that the --build isn't parsed since it is passed after example.js.
Could you try:
`nexe --build example.js`

**Comment by mturnshek** [comment]
No, that’s not working for me. Any command I run which starts with `nexe —build` results in the error referenced in my above post. 
```
Error: Entry file "" not found!

See nexe -h for usage..
```

**Comment by aral** [comment]
Same error as @mturnshek while attempting to build on Raspberry Pi 4 under Raspbian. @wilmardo: Location of `--build` option doesn't affect the outcome (either before or after the input file name produces the same error):

```
Error: python ./configure.py --dest-cpu=arm7l exited with code: 2
```

**Comment by kycfeel** [comment]
@aral I get this error after I install the master nexe by using `npm i nexe/nexe#master` as you mentioned. Is this a bug on master code?

```
root@raspberrypi:~# nexe version
internal/modules/cjs/loader.js:638
    throw err;
    ^

Error: Cannot find module './lib/options'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/usr/lib/node_modules/nexe/index.js:2:17)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
```

**Comment by aral** [comment]
It looks like there might be a regression as I just tried a new build on my Raspberry Pi 3B+ and got the following error:

```
✔ Node source extracted to: /home/pi/.nexe/12.16.2
✔ Configuring node build: --dest-cpu=arm7l
✔ Finished in 307.085s
(node:22393) UnhandledPromiseRejectionWarning: Error: python ./configure.py --dest-cpu=arm7l exited with code: 2
    at new NexeError (/home/pi/small-tech/site.js/app/node_modules/nexe/lib/compiler.js:83:28)
    at ChildProcess.<anonymous> (/home/pi/small-tech/site.js/app/node_modules/nexe/lib/compiler.js:274:28)
```

**Comment by aral** [comment]
Just to update, I just tried to build on Pi 4 and same result.

**Comment by calebboyd** [comment]
Hmm I need to release v4..You can try nexe@next to get the fix... Sorry about that.

**Comment by aral** [comment]
@calebboyd Ah, just saw your note after I added mine :)

PS. I think there might be a regression in v4 with modules required via relative paths not being found. Going to look into it now and will post what I find.

**Comment by aral** [comment]
@calebboyd Right, I’ve reproduced the regression and confirmed that it is between 3.3.2 and nexe@next: #755

**Comment by DonovanBoddy** [comment]
Is there a plan to implement this? I would like to use this library but the arm compilation is required

**Comment by kycfeel** [comment]
@DonovanBoddy For me, I'm handing it by storing a pre-compiled node source at external storage and pulls it before running the build for multi-arches. 

The complied node source will be located at `$HOME/.nexe` directory. If you are going to build binaries for ARM arch, for example, you should initially compile the node source on ARM machine, and extract the source from it (the entire `$HOME/.nexe` dir). You could compile the node source by running a `nexe` command.

After that, on your every multi-arch build, pull the complied node source and place it on your build machine's `$HOME/.nexe` directory, and run the typical `nexe` command for a build. That's it.

I'm building binaries for `s390x` and `arm64` by using this way on my x64 CI / CD setup.

**Comment by DonovanBoddy** [comment]
nice, could you maybe point me to a tutorial on how to compile the node source on the pi? I tried to just let nexe compile it on its own when it cant find the precompiled binary but it failed

**Comment by wilmardo** [comment]
I am using the same approach as @kycfeel but with QEMU and within Docker ([link to repo](https://github.com/LANsible/docker-nexe)). I use Gitlab CI to compile nexe for arm64 and amd64 (I am open to implement armhf when needed) and then push the images to [Docker Hub](https://hub.docker.com/r/lansible/nexe/tags).

I then use that image as `FROM` for my binary compilation so it has a compiled nexe available for use.
https://github.com/LANsible/docker-zigbee2mqtt/blob/master/Dockerfile

This makes cross compilation possible :)

So when you want to build the above Dockerfile for your arm64 machine on amd64 you run this:
```
// Register the QEMU binaries
# docker run --rm --privileged multiarch/qemu-user-static:register --reset
// Then build the image for arm64
# docker build . \
      --build-arg ARCHITECTURE=arm64 \
      --tag lansible/zigbee2mqtt:dev-arm64
```

I will write a blogpost on this someday but I am open for questions and improvements, just open an issue on [LANsible/docker-nexe](https://github.com/LANsible/docker-nexe)

**Comment by DonovanBoddy** [comment]
Thanks, I really appreciate it, I'll give it a bash later

**Comment by kycfeel** [comment]
@DonovanBoddy It's simple. Just get the `nexe` on your Raspberry Pi, and run `nexe` command with `--build` flag (For example: `nexe app.js -t 10.15.3 --build`). It would get the specified version of `nexe` and compile it automatically. 

Once the compile is done, you would be able to find it at `$HOME/.nexe` directory. You could also reuse this compiled node source by copying & pasting the entire directory to a different machine If you want.

**Comment by wilmardo** [comment]
>@DonovanBoddy It's simple. Just get the nexe on your Raspberry Pi, and run nexe command with --build flag (For example: nexe app.js -t 10.15.3 --build). It would get the specified version of nexe and compile it automatically.

Agree that this should work but that takes ages, it is way more efficient to use a CROSS_COMPILE setup to take advantage of more powerful hardware :)

**Comment by SamuelToh** [comment]
Hi guys,

Silly question here, apologise if there is already a solution for it. Asking again because I don't quite understand some of the responses here.

I am on `armv7l` architecture. Followed some of the potential workarounds mentioned above to build on an armhf but haven't got any luck.

What I have done is;

1. Installing the standard release for `3.3.7`, didn't work
2. `npm install git+https://github.com/nexe/nexe.git#master` resulted in -> `+ nexe@4.0.0-beta.14`
3. then do a simple `nexe -h`

I'm currently getting one of the mentioned error above;
```
internal/modules/cjs/loader.js:626
    throw err;
    ^

Error: Cannot find module './lib/options'
```

I have `nvm` installed to manage my `node` version, I wonder if this setup has anything to do with it or not. Any advise will be greatly appreciated.

**Comment by calebboyd** [comment]
@SamuelToh The nexe tool is built and published on npm, if you clone or install from the repo you have to install the project dependencies and build it (`npm i` and `npm run build` respectively)

**Comment by calebboyd** [comment]
Closing this in favor of generic #754

**Comment by zschiffelbein** [comment]
I was able to successfully build on a Raspberry Pi 4 (Took 3.5 hours). Subsequent builds run much quicker, but I'm wondering if there is a way I can take the now compiled version of node from ~/.nexe to run builds on my Mac? I attempted this by copying the node directory to my Mac but the --target flag does not seem to support building from a directory. Is there some way I can flatten the directory in to a single file like one of the assets in the V3 release?

**Comment by SamuelToh** [comment]
@calebboyd Thanks mate. Just got to try it over the weekend. Waited the goods news for 6 hours and its all good for me.

So for anybody trying to do the same thing.

- Checkout the source code (I'm using master branch)
- `npm install`
- `npm run build` (should see a lib directory added with some files)
- touch a file, `touch helloWorld.js`
- add a print stmt in there.
- then build it. `node index.js --input helloWorld.js --output myOutput`
- Wait for a bit. Some reported 3hrs+... Mine took 6hours.
- run executable `./myOutput`

**Comment by sjlongland** [comment]
Hi all,

Firstly, thanks for all the hard work put into this.  This tool does produce a nicely self-contained binary of NodeJS projects.  NodeJS isn't even required on the target machine to run the binary produced.

I'm trying to get things building on `armhf` (i.e. 32-bit ARM hard-float), the plan being to _build_ `nexe` and its binary stub, then package that into a `.deb` package which can be hosted in a repository, and pulled into any build environment where we need it instead of waiting about 1hr 15 mins for an Amazon `t4g.small` instance to compile it each time (or 3 hours on a Raspberry Pi 4).

A couple of questions though:

1. Right now I can make things _mostly_ portable; by setting `NEXE_TEMP` in my environment when doing the first run with `--build`, however the directory pointed to by `NEXE_TEMP` is huge, with references to the directory on the host in which the files were built.  Is there a way to safely slim this down a bit and maybe anonymise its source?
2. Is there a way to embed it in `NEXE_TEMP` such that `--build` is not required?  If I leave `--build` off (after running it before-hand) it seems to want to download a fresh binary down from Github, and fails, because there's no binaries for `armhf`.

I did try looking at the sources, but without doing a full-scale in-depth analysis, there wasn't any clear answer.

Regards,
Stuart Longland
